##**Machine Learning Major Project:**

###**Topic:**

Image Classification using Convolutional Neural Networks (CNNs)

###**Description:**

The project involves building an image classification system using Convolutional Neural Networks (CNNs). CNNs are widely used for image-related tasks and have achieved remarkable success in various domains. This project will allow students to gain hands-on experience with deep learning, image processing, and model development.

## Steps:

### **1. Dataset Collection:**
Gather a large dataset of labeled images relevant to the chosen domain or application. You can explore publicly available datasets like CIFAR-10, ImageNet, or create your own dataset.

### **2. Data Preprocessing:**
Preprocess the collected images by resizing them, normalizing pixel values, and splitting them into training and testing sets.

###**3. CNN Model Architecture:**
Design and implement a CNN architecture suitable for image classification. Consider popular architectures like VGGNet, ResNet, or design your own architecture based on the problem requirements.

###**4. Model Training:**
Train the CNN model using the training dataset. Experiment with different hyperparameters, such as learning rate, batch size, and optimizer, to achieve better performance.

###**5. Model Evaluation:**
Evaluate the trained model using the testing dataset. Calculate metrics such as accuracy, precision, recall, and F1-score to assess the model's performance.

###**6. Model Optimization:**
Fine-tune the model by applying techniques like data augmentation, regularization, or transfer learning to further improve performance.

###**7. Deployment and Interface:**
Build a user-friendly interface where users can input images, and the trained model will predict the image class or category. You can use web frameworks like Flask or Django for this purpose.

###**8. Documentation and Presentation:**
Document the project, including the dataset, preprocessing steps, model architecture, training process, evaluation results, and deployment instructions. Prepare a presentation to showcase the project's objectives, methodology, and outcomes.

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
# Import necessary libraries
import tensorflow as tf
from tensorflow.keras import datasets, layers, models
from tensorflow.keras.applications.vgg16 import VGG16
from tensorflow.keras.preprocessing import image
from tensorflow.keras.applications.vgg16 import preprocess_input
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import numpy as np

# Load the CIFAR-10 dataset
(train_images, train_labels), (test_images, test_labels) = datasets.cifar10.load_data()

# Normalize pixel values to be between 0 and 1
train_images, test_images = train_images / 255.0, test_images / 255.0

# Create the base model from the pre-trained model VGG16
base_model = VGG16(weights='imagenet', include_top=False, input_shape=(32, 32, 3))

# Freeze the base model
base_model.trainable = False

# Add a new top layer
x = layers.Flatten()(base_model.output)
x = layers.Dense(512, activation='relu')(x)
x = layers.Dropout(0.5)(x)
predictions = layers.Dense(10, activation='softmax')(x)

# This is the model we will train
model = models.Model(inputs=base_model.input, outputs=predictions)

# Compile the model
model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
history = model.fit(train_images, train_labels, epochs=10,
                    validation_data=(test_images, test_labels))

# Evaluate the trained model on the test set
test_loss, test_acc = model.evaluate(test_images, test_labels, verbose=2)
print(test_acc)



# Evaluate the trained model on the test set
predictions = model.predict(test_images)
y_pred = np.argmax(predictions, axis=1)
accuracy = accuracy_score(test_labels, y_pred)
precision = precision_score(test_labels, y_pred, average='macro')
recall = recall_score(test_labels, y_pred, average='macro')
f1 = f1_score(test_labels, y_pred, average='macro')

print(f'Accuracy: {accuracy:.2f}')
print(f'Precision: {precision:.2f}')
print(f'Recall: {recall:.2f}')
print(f'F1-score: {f1:.2f}')

# Fine-tune the model using data augmentation and transfer learning
# Create a new base model from the pre-trained VGG16 model
base_model = VGG16(weights='imagenet', include_top=False, input_shape=(32, 32, 3))

# Freeze all layers in the base model except the last block of layers
for layer in base_model.layers[:-4]:
    layer.trainable = False

# Add a new top layer to the model
x = layers.Flatten()(base_model.output)
x = layers.Dense(512, activation='relu')(x)
x = layers.Dropout(0.5)(x)
predictions = layers.Dense(10, activation='softmax')(x)

# This is the new model we will fine-tune
model = models.Model(inputs=base_model.input, outputs=predictions)

# Compile the model
model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.0001),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Create an ImageDataGenerator for data augmentation
datagen = ImageDataGenerator(
    rotation_range=20,
    width_shift_range=0.2,
    height_shift_range=0.2,
    horizontal_flip=True)

# Fit the data generator on the training data
datagen.fit(train_images)

# Fine-tune the model using the augmented data
history = model.fit(datagen.flow(train_images, train_labels),
                    epochs=10,
                    validation_data=(test_images, test_labels))

# Evaluate the fine-tuned model on the test set
test_loss, test_acc = model.evaluate(test_images, test_labels, verbose=2)
print(test_acc)

# Save the trained model to an .h5 file
model.save('image_classification_model.h5')

58889256/58889256 [==============================] - 4s 0us/step
Epoch 1/10
1563/1563 [==============================] - 27s 11ms/step - loss: 1.4368 - accuracy: 0.4954 - val_loss: 1.2665 - val_accuracy: 0.5530
Epoch 2/10
1563/1563 [==============================] - 16s 10ms/step - loss: 1.2699 - accuracy: 0.5537 - val_loss: 1.2071 - val_accuracy: 0.5735
Epoch 3/10
1563/1563 [==============================] - 17s 11ms/step - loss: 1.2126 - accuracy: 0.5748 - val_loss: 1.1769 - val_accuracy: 0.5846
Epoch 4/10
1563/1563 [==============================] - 17s 11ms/step - loss: 1.1781 - accuracy: 0.5868 - val_loss: 1.1486 - val_accuracy: 0.5911
Epoch 5/10
1563/1563 [==============================] - 16s 10ms/step - loss: 1.1523 - accuracy: 0.5943 - val_loss: 1.1359 - val_accuracy: 0.5993
Epoch 6/10
1563/1563 [==============================] - 16s 10ms/step - loss: 1.1295 - accuracy: 0.6032 - val_loss: 1.1247 - val_accuracy: 0.6056
Epoch 7/10
1563/1563 [==============================] - 16s

**Project Documentation: Image Classification with Transfer Learning
This documentation provides a comprehensive overview of the image classification project, encompassing dataset details, preprocessing steps, model architecture, training process, evaluation outcomes, and deployment instructions.**

1.	Dataset: CIFAR-10
- The CIFAR-10 dataset consists of 60,000 color images, each sized 32x32 pixels, distributed across 10 distinct classes.
- The dataset is divided into a training set containing 50,000 images and a test set containing 10,000 images.

2.	Preprocessing:
- Preprocessing involves pixel normalization, which scales pixel values within the range of [0, 1].
  
3.	Model Architecture:
- The foundational model is VGG16, pretrained on the ImageNet dataset. It excludes the fully connected layers.
- Additional layers include: Flatten, Dense (512 units, ReLU activation), Dropout (rate = 0.5), Dense (10 units, softmax activation).

4.	Training Process:
- Initialization involves loading pretrained VGG16 weights and keeping them frozen.
- The complete model is compiled using the Adam optimizer and sparse categorical cross-entropy loss.
- Initial training spans 10 epochs on the training set, with validation data for monitoring progress.
- Subsequent fine-tuning involves data augmentation.

5.	Evaluation Results:
- After each training phase, the model's performance is evaluated on the test set.
- The test accuracy achieved after the initial training phase is approximately 61.71%.
- After fine-tuning, the accuracy decreases to around 9.84% due to significant modifications in the model architecture and the freezing of certain layers.
  - For the initial model:
  - Precision: 0.62
  - Recall: 0.62
  - F1-score: 0.61

6.	Deployment Instructions:
- To deploy the model for inference, it's essential to save the model architecture and weights.
- Loading the saved model and employing it for making predictions are the fundamental deployment steps.
- Make sure to have the requisite libraries (TensorFlow, scikit-learn) installed for a seamless deployment experience.
- The deployment avenues include cloud platforms, mobile applications, and web services based on specific project requirements.

Output Summary:
 - Initial model accuracy on the test set: ~61.71%
 - Accuracy after fine-tuning: ~9.84%
 - For the initial model:
 - Precision: 0.62
 - Recall: 0.62
 - F1-score: 0.61
